File: modules/androidkit/src/Paint.cpp
/*
 * Copyright 2021 Google Inc.
 *
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

#include <jni.h>

#include "include/core/SkColorFilter.h"
#include "include/core/SkPaint.h"
#include "include/core/SkShader.h"
#include "include/effects/SkImageFilters.h"

namespace {

static jlong Paint_Create(JNIEnv*, jobject) {
    return reinterpret_cast<jlong>(new SkPaint);
}

static void Paint_Release(JNIEnv*, jobject, jlong native_paint) {
    delete reinterpret_cast<SkPaint*>(native_paint);
}

static void Paint_SetColor(JNIEnv*, jobject, jlong native_paint,
                           float r, float g, float b, float a) {
    if (auto* paint = reinterpret_cast<SkPaint*>(native_paint)) {
        paint->setColor4f({r, g, b, a});
    }
}

static void Paint_SetStroke(JNIEnv*, jobject, jlong native_paint, jboolean stroke) {
    if (auto* paint = reinterpret_cast<SkPaint*>(native_paint)) {
        paint->setStroke(stroke);
    }
}

static void Paint_SetStrokeWidth(JNIEnv*, jobject, jlong native_paint, jfloat width) {
    if (auto* paint = reinterpret_cast<SkPaint*>(native_paint)) {
        paint->setStrokeWidth(width);
    }
}

static void Paint_SetStrokeCap(JNIEnv* env, jobject, jlong native_paint, jint native_cap) {
    if (auto* paint = reinterpret_cast<SkPaint*>(native_paint)) {
        switch (native_cap)
        {
        case 0:
            paint->setStrokeCap(SkPaint::kButt_Cap);
            break;
        case 1:
            paint->setStrokeCap(SkPaint::kRound_Cap);
            break;
        case 2:
            paint->setStrokeCap(SkPaint::kSquare_Cap);
            break;
        default:
            break;
        }
    }
}

static void Paint_SetStrokeJoin(JNIEnv* env, jobject, jlong native_paint, jint native_join) {
    if (auto* paint = reinterpret_cast<SkPaint*>(native_paint)) {
        switch (native_join)
        {
        case 0:
            paint->setStrokeJoin(SkPaint::kMiter_Join);
            break;
        case 1:
            paint->setStrokeJoin(SkPaint::kRound_Join);
            break;
        case 2:
            paint->setStrokeJoin(SkPaint::kBevel_Join);
            break;
        default:
            break;
        }
    }
}

static void Paint_SetStrokeMiter(JNIEnv* env, jobject, jlong native_paint, jfloat limit) {
    if (auto* paint = reinterpret_cast<SkPaint*>(native_paint)) {
        paint->setStrokeMiter(limit);
    }
}

static void Paint_SetColorFilter(JNIEnv*, jobject, jlong native_paint, jlong native_cf) {
    if (auto* paint = reinterpret_cast<SkPaint*>(native_paint)) {
        paint->setColorFilter(sk_ref_sp(reinterpret_cast<SkColorFilter*>(native_cf)));
    }
}

static void Paint_SetShader(JNIEnv*, jobject, jlong native_paint, jlong native_shader) {
    if (auto* paint = reinterpret_cast<SkPaint*>(native_paint)) {
        paint->setShader(sk_ref_sp(reinterpret_cast<SkShader*>(native_shader)));
    }
}

static void Paint_SetImageFilter(JNIEnv*, jobject, jlong native_paint, jlong native_filter) {
    if (auto* paint = reinterpret_cast<SkPaint*>(native_paint)) {
        paint->setImageFilter(sk_ref_sp(reinterpret_cast<SkImageFilter*>(native_filter)));
    }
}

}  // namespace

int register_androidkit_Paint(JNIEnv* env) {
    static const JNINativeMethod methods[] = {
        {"nCreate"         , "()J"     , reinterpret_cast<void*>(Paint_Create)},
        {"nRelease"        , "(J)V"    , reinterpret_cast<void*>(Paint_Release)},
        {"nSetColor"       , "(JFFFF)V", reinterpret_cast<void*>(Paint_SetColor)},
        {"nSetStroke"      , "(JZ)V"   , reinterpret_cast<void*>(Paint_SetStroke)},
        {"nSetStrokeWidth" , "(JF)V"   , reinterpret_cast<void*>(Paint_SetStrokeWidth)},
        {"nSetStrokeCap"   , "(JI)V"   , reinterpret_cast<void*>(Paint_SetStrokeCap)},
        {"nSetStrokeJoin"  , "(JI)V"   , reinterpret_cast<void*>(Paint_SetStrokeJoin)},
        {"nSetStrokeMiter" , "(JF)V"   , reinterpret_cast<void*>(Paint_SetStrokeMiter)},
        {"nSetColorFilter" , "(JJ)V"   , reinterpret_cast<void*>(Paint_SetColorFilter)},
        {"nSetShader"      , "(JJ)V"   , reinterpret_cast<void*>(Paint_SetShader)},
        {"nSetImageFilter" , "(JJ)V"   , reinterpret_cast<void*>(Paint_SetImageFilter)},
    };

    const auto clazz = env->FindClass("org/skia/androidkit/Paint");
    return clazz
        ? env->RegisterNatives(clazz, methods, SK_ARRAY_COUNT(methods))
        : JNI_ERR;
}

File: platform_tools/android/apps/AndroidKit/src/main/java/org/skia/androidkit/Paint.java
/*
 * Copyright 2021 Google Inc.
 *
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

package org.skia.androidkit;

import android.support.annotation.Nullable;

public class Paint {
    private long mNativeInstance;

    public Paint() {
        mNativeInstance = nCreate();
    }

    public Paint setColor(Color c) {
        nSetColor(mNativeInstance, c.r(), c.g(), c.b(), c.a());
        return this;
    }

    public Paint setColor(float r, float g, float b, float a) {
        nSetColor(mNativeInstance, r, g, b, a);
        return this;
    }

    public Paint setColorFilter(@Nullable ColorFilter filter) {
        nSetColorFilter(mNativeInstance, filter != null ? filter.getNativeInstance() : 0);
        return this;
    }

    public Paint setShader(@Nullable Shader shader) {
        nSetShader(mNativeInstance, shader != null ? shader.getNativeInstance() : 0);
        return this;
    }

    public Paint setImageFilter(@Nullable ImageFilter filter) {
        nSetImageFilter(mNativeInstance, filter != null ? filter.getNativeInstance() : 0);
        return this;
    }

    public Paint setStroke(boolean stroke) {
        nSetStroke(mNativeInstance, stroke);
        return this;
    }

    public Paint setStrokeWidth(float w) {
        nSetStrokeWidth(mNativeInstance, w);
        return this;
    }

    public enum Cap {
        BUTT    (0),
        ROUND   (1),
        SQUARE  (2);

        private Cap(int nativeInt) {
            this.nativeInt = nativeInt;
        }
        final int nativeInt;
    }
    public Paint setStrokeCap(Cap cap) {
        nSetStrokeCap(mNativeInstance, cap.nativeInt);
        return this;
    }


    public enum Join {
        MITER   (0),
        ROUND   (1),
        BEVEL   (2);

        private Join(int nativeInt) {
            this.nativeInt = nativeInt;
        }
        final int nativeInt;
    }
    public Paint setStrokeJoin(Join join) {
        nSetStrokeJoin(mNativeInstance, join.nativeInt);
        return this;
    }

    public Paint setStrokeMiter(float limit) {
        nSetStrokeMiter(mNativeInstance, limit);
        return this;
    }

    /**
     * Releases any resources associated with this Paint.
     */
    public void release() {
        nRelease(mNativeInstance);
        mNativeInstance = 0;
    }

    @Override
    protected void finalize() throws Throwable {
        release();
    }

    // package private
    long getNativeInstance() { return mNativeInstance; }

    private static native long nCreate();
    private static native void nRelease(long nativeInstance);

    private static native void nSetColor(long nativeInstance, float r, float g, float b, float a);
    private static native void nSetStroke(long nativeInstance, boolean stroke);
    private static native void nSetStrokeWidth(long nativeInstance, float w);
    private static native void nSetStrokeCap(long nativeInstance, int native_cap);
    private static native void nSetStrokeJoin(long nativeInstance, int native_join);
    private static native void nSetStrokeMiter(long nativeInstance, float limit);
    private static native void nSetColorFilter(long nativeInstance, long nativeCF);
    private static native void nSetShader(long nativeInstance, long nativeShader);
    private static native void nSetImageFilter(long nativeInstance, long nativeFilter);
}
